### PR TITLE
Feat/aave v3

### PR DESCRIPTION
**Pull request by majkic99**

### CODE REVIEW

File: contracts/actions/aaveV3/helpers/AaveV3Helper.sol [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import "./OptimismAaveV3Addresses.sol";
+
+import "../../../interfaces/aaveV3/IL2PoolV3.sol";
+import "../../../interfaces/aaveV3/IAaveProtocolDataProvider.sol";
+import "../../../interfaces/aaveV3/IPoolAddressesProvider.sol";
+
+/// @title Utility functions and data used in AaveV3 actions
+contract AaveV3Helper is OptimismAaveV3Addresses {
+    /// TODO: Change this later
+    uint16 public constant AAVE_REFERRAL_CODE = 64;
```
Comment (yippee-ki-yay) [inline contracts/actions/aaveV3/helpers/AaveV3Helper.sol:13]: Should be checked before deployment

File: contracts/actions/aaveV3/helpers/AaveV3Helper.sol [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// @notice Fetch the data provider for the specified market
+    function getDataProvider(address _market) internal view returns (IAaveProtocolDataProvider) {
+        return
+            IAaveProtocolDataProvider(
+                IPoolAddressesProvider(_market).getPoolDataProvider()
+            );
+    }
+
+    function boolToBytes(bool x) internal pure returns (bytes1 r) {
```
Comment (yippee-ki-yay) [inline contracts/actions/aaveV3/helpers/AaveV3Helper.sol:32]: While the gas savings will probably be really low, there is no need here to switch back and forth to uint8, we can compare values with bytes1, for instance the following two function can be written as:

```
function boolToBytes(bool x) internal pure returns (bytes1) {
        return x == true ? bytes1(0x01) : bytes1(0x00);
}

    function bytesToBool(bytes1 x) internal pure returns (bool) {
        return x != bytes1(0x00);
    }
```

File: contracts/actions/aaveV3/AaveV3SetEMode.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /// @inheritdoc ActionBase
+    function actionType() public pure virtual override returns (uint8) {
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    /// @notice User deposits tokens to the Aave protocol
```
Comment (yippee-ki-yay) [inline contracts/actions/aaveV3/AaveV3SetEMode.sol:55]: Copy paste source of all evil

File: contracts/actions/aaveV3/AaveV3ClaimRewards.sol [hunk 4]
```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.10;
+
+import "../../utils/TokenUtils.sol";
+import "../ActionBase.sol";
+import "./helpers/AaveV3Helper.sol";
+import "../../interfaces/aaveV3/IRewardsController.sol";
+
+/// @title Supply a token to an Aave market
```
Comment (yippee-ki-yay) [inline contracts/actions/aaveV3/AaveV3ClaimRewards.sol:10]: Comments ser

File: contracts/actions/aaveV3/AaveV3ClaimRewards.sol [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        bytes memory logData = abi.encode(params, amountReceived);
+        return (amountReceived, logData);
+    }
+
+    function parseInputs(bytes memory _callData) public pure returns (Params memory params) {
+        params = abi.decode(_callData, (Params));
+    }
+
+    function encodeInputs(Params memory params) public pure returns (bytes memory encodedInput) {
+        require(params.assetsLength == params.assets.length);
```
Comment (yippee-ki-yay) [inline contracts/actions/aaveV3/AaveV3ClaimRewards.sol:83]: Add the same check to CollateralSwitch?

File: test/aaveV3/aave-atoken-payback.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+const {
+    getProxy, balanceOf, setBalance, approve, redeploy, takeSnapshot, revertToSnapshot,
+} = require('../utils');
+
+describe('AaveV3-ATokenPayback-L2', function () {
+    this.timeout(150000);
+
+    let senderAcc; let proxy; let pool; let snapshotId;
+    const WETH_ADDRESS = '0x4200000000000000000000000000000000000006';
```
Comment (yippee-ki-yay) [inline test/aaveV3/aave-atoken-payback.js:16]: utils.js addrs array has this based on the network

Comment (yippee-ki-yay) [reply test/aaveV3/aave-atoken-payback.js:16]: don't hardcode optimism info so we can use for other L2 later on

**Review by yippee-ki-yay — APPROVED** [review]
In general the code is good a few small improvements can be added and some comments fixed

File: contracts/actions/aaveV3/AaveV3SwapBorrowRateMode.sol [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import "../ActionBase.sol";
+import "./helpers/AaveV3Helper.sol";
+
+/// @title Swaps proxy positions borrow rate mode between stable and variable.
+contract AaveV3SwapBorrowRateMode is ActionBase, AaveV3Helper {
+    using TokenUtils for address;
+    struct Params {
+        address asset;
+        uint256 rateMode;
+        bool useDefaultMarket;
```
Comment (kjaredic) [inline contracts/actions/aaveV3/AaveV3SwapBorrowRateMode.sol:15]: this uses a slot, maybe have address(0) mean default market? the decode function could infer this from encodedInput.length

**Review by kjaredic — COMMENTED** [review]
code looks good

**Review by kjaredic — APPROVED** [review]
lgtm

**Pull request merged**